GetDKAN/dkan
# Form widget based on JSON Schema

## Test steps
- [x] Enable the json_form_widget module.
- [x] Go to Structure -> Content types -> Data -> Manage form display, select the widget "JSON Form" for the field "JSON Metadata", make the schema be "dataset" and save the configuration.
- [x] Create a new data node, the form should load correctly with fields for the different elements from the schema defined. Once you add values and save, a new dataset should be created correctly with JSON saved in the node (check the node page to confirm).
- [x] Edit an existing dataset, the form should load correctly with all the information from the node, you should be able to edit some values and save the form without errors. 
- Confirm the following is true on the forms:
  - [ ] If the schema configured for the field does not exist, then the form is not shown.
  - [ ] If the schema UI does not exist, then the form is still shown based just in the JSON Schema.
  - [x] You can add multiple Categories and Tags to a node.
  - [x] You can add multiple distributions to a dataset.
  - [x] You can add multiple fields to the "Schema" of a distribution.
- The schema UI should be applied correctly, check:
  - [x] You can make fields to be hidden by adding:
```
"@type": {
    "ui:options": {
      "widget": "hidden"
    }
  },
```

  - [x] You can change the widget of a field to be a textarea and configure the amount of rows and cols displayed by adding:
```
"spatial": {
    "ui:options": {
      "widget": "textarea",
      "rows": 5,
      "cols": 60,
    }
  },
```

  - [x] You can disable a field by adding:
```
"accessLevel": {
    "ui:options": {
      "disabled": true
    }
  },
```

  - [x] You can add a placeholder to fields by adding:
```
"modified": {
    "ui:options": {
      "placeholder": "YYYY-MM-DD"
    }
  },
```

  - [x] You can override the field description/help text by adding:
```
"field_name": {
    "ui:options": {
      "description": "Test help text.",
    }
  },
```

  - [x] You can override the field title by adding:
```
"description": {
    "ui:options": {
      "title": "More descriptive title"
    }
  },
```
  - [x] You can use the dkan_uuid widget in a field to pre-fill it with a valid UUID and make it invisible:
```
"identifier": {	
    "ui:options": {	
      "widget": "dkan_uuid"	
    }	
  },
```
  - [x] You can override array fields and its subfields:
```
"references": {
    "ui:options": {
      "title": "Better title.",
    },
    "items": {
      "ui:options": {
        "placeholder": "http://"
      }
    }
  },
```
  - [x] You can override object fields and its subfields:
```
"publisher": {
    "ui:options": {
      "description": "Better description.",
    },
    "properties": {
      "@type": {
        "ui:options": {
          "widget": "hidden"
        }
      }
    }
  },
```
